cpp/nacjd.cc

```diff
@@ -0,0 +1,361 @@
+#include "Downloader.h"
+#include "FileUtil.h"
+#include "HttpHeader.h"
+#include "JSON.h"
+#include "MARC.h"
+#include "UrlUtil.h"
+#include "util.h"
+#include "UBTools.h"
+
+
+namespace {
+
+
+[[noreturn]] void Usage() {
+    ::Usage("marc_title_in_file new_marc_title_out_file\n"
+            "where marc_title_in_file contains also icpsr records (001 or 035a)\n"
+            "these records are not processed any more\n"
+            "new_marc_title_out_file contains all icpsr records not contained in input file.");
+}
+
+
+std::vector<std::string> ids_website;
+const unsigned int TIMEOUT_IN_SECONDS(15);
+const std::string NACJD_TITLES(UBTools::GetTuelibPath() + "nacjd_titles.html");
+const std::string NACJD_NEW_TITLES_JSON(UBTools::GetTuelibPath() + "nacjd_new_titles.json");
+
+
+bool ContainsValue(const std::map<std::string, std::string> &map, const std::string &search_value) {
+    for (const auto &[key, val] : map) {
+        if (val == search_value)
+            return true;
+    }
+    return false;
+}
+
+
+enum PreviousChar { ID_START, FIRST_DOUBLE_QUOTE_SEEN, CAP_I_SEEN, CAP_D_SEEN, SECOND_DOUBLE_QUOTE_SEEN, COLON_SEEN };
+
+
+void HandleChar(const char c) {
+    static std::string current_id_website;
+    static PreviousChar state(ID_START);
+    if (state == ID_START) {
+        if (c == '"')
+            state = FIRST_DOUBLE_QUOTE_SEEN;
+    } else if (state == FIRST_DOUBLE_QUOTE_SEEN) {
+        if (c == 'I')
+            state = CAP_I_SEEN;
+        else
+            state = ID_START;
+    } else if (state == CAP_I_SEEN) {
+        if (c == 'D')
+            state = CAP_D_SEEN;
+        else
+            state = ID_START;
+    } else if (state == CAP_D_SEEN) {
+        if (c == '"')
+            state = SECOND_DOUBLE_QUOTE_SEEN;
+        else
+            state = ID_START;
+    } else if (state == SECOND_DOUBLE_QUOTE_SEEN) {
+        if (c == ':')
+            state = COLON_SEEN;
+        else
+            state = ID_START;
+    } else if (state == COLON_SEEN) {
+        if (c == ',') {
+            ids_website.push_back(current_id_website);
+            current_id_website.clear();
+            state = ID_START;
+        } else
+            current_id_website += c;
+    }
+}
+
+
+bool DownloadID(std::ofstream &json_new_titles, const std::string &id, const bool use_separator) {
+    const std::string DOWNLOAD_URL("https://pcms.icpsr.umich.edu/pcms/api/1.0/studies/" + id + "/dats?page=https://www.icpsr.umich.edu/web/NACJD/studies/" + id + "/export&user=");
+
+    Downloader downloader(DOWNLOAD_URL, Downloader::Params(), TIMEOUT_IN_SECONDS * 1000);
+    if (downloader.anErrorOccurred()) {
+        LOG_WARNING("Error while downloading data for id " + id + ": " + downloader.getLastErrorMessage());
+        return false;
+    }
+
+     // Check for rate limiting and error status codes:
+    const HttpHeader http_header(downloader.getMessageHeader());
+    if (http_header.getStatusCode() == 429)
+        LOG_ERROR("we got rate limited!");
```

Comment:
Can we really get here? I'd suppose that downloader.anErrorOccurred() would return true in this case.

Reply:
will remove this, maybe can also be remove in crossref_downloader then.